src/openms/source/ANALYSIS/MAPMATCHING/FeatureGroupingAlgorithm.cpp

```diff
@@ -170,6 +191,94 @@ namespace OpenMS
     }
   }
 
+  void FeatureGroupingAlgorithm::annotateIonIdentityNetworks(ConsensusMap& out) const
+  {
+    // bipartite graph with ConsensusFeature indexes and Groups from Features
+    // Vertexes contain uid (index/Group) and is_feature (bool)
+    // e.g. 
+    // ConsensusFeature0 contains Feature (Group = 1) and Feature (Group = 2)
+    // ConsensusFeature1 contains Feature (Group = 2) and Feature (Group = 3)
+    // ConsensusFeature2 contains Feature (Group = 4) and Feature (Group = 5)
+    // graph looks like (Vertex <--> Vertex; component_number)
+    // 0, true <--> 1, false; 0
+    // 0, true <--> 2, false; 0
+    // 1, true <--> 2, false; 0
+    // 1, true <--> 3, false; 0
+    // 2, true <--> 4, false; 1
+    // 2, true <--> 5, false; 1
+    // Total number of components: 2
+    UndirectedOSMIdGraph g;
+
+    for (size_t i = 0; i < out.size(); i++)
+    {
+      out[i].setMetaValue(Constants::UserParam::IIMN_ROW_ID, i+1);
+      if (!out[i].metaValueExists(Constants::UserParam::LINKED_GROUPS)) continue;
+      auto feature_vertex = add_vertex(VertexLabel(String(i), true), g);
+      for (const auto& group: out[i].getMetaValue(Constants::UserParam::LINKED_GROUPS).toStringList())
+      {
+        bool already_set = false;
+        for (auto i : boost::make_iterator_range(vertices(g)))
+        {
+            if ((group == g[i].uid) && (!g[i].is_feature)) 
+            {
+              boost::add_edge(feature_vertex, i, g);
+              already_set = true;
+              break;
+            }
+        }
+        if (!already_set)
+        {
+        boost::add_edge(feature_vertex, add_vertex(VertexLabel(group, false), g), g);
+        }
+      }
+    }
+
+    std::vector<int> components (boost::num_vertices (g));
+    boost::connected_components (g, &components[0]);
+
+    // annotate network number and create a map with feature ID and partner IDs
+    // partner feature vertexes are connected via a group vertex
+    map<size_t, set<size_t>> partner_map;
```

Comment:
Cant you allocate a vector as big as the feature map? and then access by index with the uid?

Reply:
so do more or less all of them have partners or would we gain something from using the map (e.g., save memory because many don't have partners)

Reply:
Most of the features don't have partners, so this could be more efficient.

Reply:
then I would keep it as is (or change to) unordered_map